## 기계학습 실습
### Titanic Data

먼저 Titanic Data 중 train.csv 데이터를 불러온다.<br>
pandas 모듈의 read_csv(path) 함수를 이용해 불러올 수 있다.

In [1]:
import pandas as pd
train_path = "C:/Download/train.csv"
titanic = pd.read_csv(train_path)
titanic

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
5              6         0       3   
6              7         0       1   
7              8         0       3   
8              9         1       3   
9             10         1       2   
10            11         1       3   
11            12         1       1   
12            13         0       3   
13            14         0       3   
14            15         0       3   
15            16         1       2   
16            17         0       3   
17            18         1       2   
18            19         0       3   
19            20         1       3   
20            21         0       2   
21            22         1       2   
22            23         1       3   
23            24         1       1   
24            25         0       3   
25            26         1       3   
26            27         0       3   
27            28         0       1   
28            29         1       3   
29            30         0       3   
..           ...       ...     ...   
861          862         0       2   
862          863         1       1   
863          864         0       3   
864          865         0       2   
865          866         1       2   
866          867         1       2   
867          868         0       1   
868          869         0       3   
869          870         1       3   
870          871         0       3   
871          872         1       1   
872          873         0       1   
873          874         0       3   
874          875         1       2   
875          876         1       3   
876          877         0       3   
877          878         0       3   
878          879         0       3   
879          880         1       1   
880          881         1       2   
881          882         0       3   
882          883         0       3   
883          884         0       2   
884          885         0       3   
885          886         0       3   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
5                                     Moran, Mr. James    male   NaN      0   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
7                       Palsson, Master. Gosta Leonard    male   2.0      3   
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
12                      Saundercock, Mr. William Henry    male  20.0      0   
13                         Andersson, Mr. Anders Johan    male  39.0      1   
14                Vestrom, Miss. Hulda Amanda Adolfina  female  14.0      0   
15                    Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
16                                Rice, Master. Eugene    male   2.0      4   
17                        Williams, Mr. Charles Eugene    male   NaN      0   
18   Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0      1   
19                             Masselmani, Mrs. Fatima  female   NaN

다음으로 데이터를 분석하는 데에 필요없는 열들을 삭제한다.<br>
Id code나 이름, ticket과 같은 overfitting을 유발할 만한 정보는 쓸모 없으므로 삭제한다.<br>
행이나 열을 통째로 삭제할 땐 drop 메서드를 쓰면 된다.<br>
열 정보를 삭제하려면 columns = '열 이름'의 형태로 써 주어야 한다.

In [2]:
titanic = titanic.drop(columns = 'PassengerId')

In [3]:
titanic = titanic.drop(columns = 'Name')

In [4]:
titanic = titanic.drop(columns = 'Ticket')

세 개의 열을 삭제한 후의 정보이다.<br>
간단하게 각 열 별로 몇 개의 valid한 값이 존재하는지 확인하려면 count()함수를 쓰면 된다.

In [5]:
titanic.count()

Survived    891
Pclass      891
Sex         891
Age         714
SibSp       891
Parch       891
Fare        891
Cabin       204
Embarked    889
dtype: int64

확인해보니 Cabin 열의 값의 개수가 204개밖에 되지 않는다.<br>
이는 결측값의 수가 매우 많다는 의미이므로, Cabin 열도 삭제해주자.

In [6]:
titanic = titanic.drop(columns = 'Cabin')

이제 남은 열들에서 결측값이 있는 행을 삭제해준다.<br>
dropna 함수는 결측값이 있는 행이나 열을 통째로 삭제해준다.<br>
이 때 axis의 값을 0으로 설정하면 행, 1로 설정하면 열을 나타낸다.<br>
how의 값을 'any'로 설정하면 해당 행에 하나라도 결측값이 있을 경우 해당 행을 삭제하며,<br>
'all'로 설정하면 해당 행의 모든 값이 결측값일 경우 행을 삭제한다.

In [7]:
titanic = titanic.dropna(axis=0, how='any')

In [8]:
titanic.count()

Survived    712
Pclass      712
Sex         712
Age         712
SibSp       712
Parch       712
Fare        712
Embarked    712
dtype: int64

In [9]:
titanic

Survived  Pclass     Sex   Age  SibSp  Parch      Fare Embarked
0           0       3    male  22.0      1      0    7.2500        S
1           1       1  female  38.0      1      0   71.2833        C
2           1       3  female  26.0      0      0    7.9250        S
3           1       1  female  35.0      1      0   53.1000        S
4           0       3    male  35.0      0      0    8.0500        S
6           0       1    male  54.0      0      0   51.8625        S
7           0       3    male   2.0      3      1   21.0750        S
8           1       3  female  27.0      0      2   11.1333        S
9           1       2  female  14.0      1      0   30.0708        C
10          1       3  female   4.0      1      1   16.7000        S
11          1       1  female  58.0      0      0   26.5500        S
12          0       3    male  20.0      0      0    8.0500        S
13          0       3    male  39.0      1      5   31.2750        S
14          0       3  female  14.0      0      0    7.8542        S
15          1       2  female  55.0      0      0   16.0000        S
16          0       3    male   2.0      4      1   29.1250        Q
18          0       3  female  31.0      1      0   18.0000        S
20          0       2    male  35.0      0      0   26.0000        S
21          1       2    male  34.0      0      0   13.0000        S
22          1       3  female  15.0      0      0    8.0292        Q
23          1       1    male  28.0      0      0   35.5000        S
24          0       3  female   8.0      3      1   21.0750        S
25          1       3  female  38.0      1      5   31.3875        S
27          0       1    male  19.0      3      2  263.0000        S
30          0       1    male  40.0      0      0   27.7208        C
33          0       2    male  66.0      0      0   10.5000        S
34          0       1    male  28.0      1      0   82.1708        C
35          0       1    male  42.0      1      0   52.0000        S
37          0       3    male  21.0      0      0    8.0500        S
38          0       3  female  18.0      2      0   18.0000        S
..        ...     ...     ...   ...    ...    ...       ...      ...
856         1       1  female  45.0      1      1  164.8667        S
857         1       1    male  51.0      0      0   26.5500        S
858         1       3  female  24.0      0      3   19.2583        C
860         0       3    male  41.0      2      0   14.1083        S
861         0       2    male  21.0      1      0   11.5000        S
862         1       1  female  48.0      0      0   25.9292        S
864         0       2    male  24.0      0      0   13.0000        S
865         1       2  female  42.0      0      0   13.0000        S
866         1       2  female  27.0      1      0   13.8583        C
867         0       1    male  31.0      0      0   50.4958        S
869         1       3    male   4.0      1      1   11.1333        S
870         0       3    male  26.0      0      0    7.8958        S
871         1       1  female  47.0      1      1   52.5542        S
872         0       1    male  33.0      0      0    5.0000        S
873         0       3    male  47.0      0      0    9.0000        S
874         1       2  female  28.0      1      0   24.0000        C
875         1       3  female  15.0      0      0    7.2250        C
876         0       3    male  20.0      0      0    9.8458        S
877         0       3    male  19.0      0      0    7.8958        S
879         1       1  female  56.0      0      1   83.1583        C
880         1       2  female  25.0      0      1   26.0000        S
881         0       3    male  33.0      0      0    7.8958        S
882         0       3  female  22.0      0      0   10.5167        S
883         0       2    male  28.0      0      0   10.5000        S
884         0       3    male  25.0      0      0    7.0500        S
885         0       3  female  39.0      0      5   29.1250        Q
886

결측값까지 정리된 titanic 데이터이다.<br>
8개의 열(feature)과 712개의 행(instance)를 가짐을 알 수 있다.<br>
***
이제 전체 데이터를 training set과 validation set으로 분리해보자.<br>
sample()함수는 dataframe에서 행을 기준으로 한 random sample을 반환한다.<br>
frac 값은 어느 정도의 양을 분리해낼 것인지를 나타낸다.<br>
여기서는 training set을 0.8, 즉 80%만큼 분리해내보자.

In [10]:
train = titanic.sample(frac=0.8)

다음과 같이 trian에 들어갈 instance 570개가 무작위로 선택된 것을 볼 수 있다.

In [11]:
train

Survived  Pclass     Sex    Age  SibSp  Parch      Fare Embarked
110         0       1    male  47.00      0      0   52.0000        S
814         0       3    male  30.50      0      0    8.0500        S
328         1       3  female  31.00      1      1   20.5250        S
882         0       3  female  22.00      0      0   10.5167        S
582         0       2    male  54.00      0      0   26.0000        S
569         1       3    male  32.00      0      0    7.8542        S
399         1       2  female  28.00      0      0   12.6500        S
489         1       3    male   9.00      1      1   15.9000        S
586         0       2    male  47.00      0      0   15.0000        S
434         0       1    male  50.00      1      0   55.9000        S
838         1       3    male  32.00      0      0   56.4958        S
590         0       3    male  35.00      0      0    7.1250        S
377         0       1    male  27.00      0      2  211.5000        C
670         1       2  female  40.00      1      1   39.0000        S
305         1       1    male   0.92      1      2  151.5500        S
329         1       1  female  16.00      0      1   57.9792        C
311         1       1  female  18.00      2      2  262.3750        C
133         1       2  female  29.00      1      0   26.0000        S
606         0       3    male  30.00      0      0    7.8958        S
664         1       3    male  20.00      1      0    7.9250        S
106         1       3  female  21.00      0      0    7.6500        S
414         1       3    male  44.00      0      0    7.9250        S
856         1       1  female  45.00      1      1  164.8667        S
74          1       3    male  32.00      0      0   56.4958        S
446         1       2  female  13.00      0      1   19.5000        S
58          1       2  female   5.00      1      2   27.7500        S
157         0       3    male  30.00      0      0    8.0500        S
14          0       3  female  14.00      0      0    7.8542        S
515         0       1    male  47.00      0      0   34.0208        S
12          0       3    male  20.00      0      0    8.0500        S
..        ...     ...     ...    ...    ...    ...       ...      ...
679         1       1    male  36.00      0      1  512.3292        C
234         0       2    male  24.00      0      0   10.5000        S
238         0       2    male  19.00      0      0   10.5000        S
443         1       2  female  28.00      0      0   13.0000        S
875         1       3  female  15.00      0      0    7.2250        C
865         1       2  female  42.00      0      0   13.0000        S
758         0       3    male  34.00      0      0    8.0500        S
597         0       3    male  49.00      0      0    0.0000        S
871         1       1  female  47.00      1      1   52.5542        S
471         0       3    male  38.00      0      0    8.6625        S
528         0       3    male  39.00      0      0    7.9250        S
286         1       3    male  30.00      0      0    9.5000        S
224         1       1    male  38.00      1      0   90.0000        S
188         0       3    male  40.00      1      1   15.5000        Q
242         0       2    male  29.00      0      0   10.5000        S
809         1       1  female  33.00      1      0   53.1000        S
39          1       3  female  14.00      1      0   11.2417        C
2           1       3  female  26.00      0      0    7.9250        S
280         0       3    male  65.00      0      0    7.7500        Q
890         0       3    male  32.00      0      0    7.7500        Q
447         1       1    male  34.00      0      0   26.5500        S
233         1       3  female   5.00      4      2   31.3875        S
544         0       1    male  50.00      1      0  106.4250        C
81          1       3    male  29.00      0      0    9.5000        S
673         1       2    male  31.00      0      0   13.0000        S
858         1  

다음은 validation set을 만들어보자.<br>
이미 training set을 만들었기 때문에 전체 titanic 데이터에서 train set을 제외하면 validation set을 쉽게 만들 수 있다.<br>
drop 함수를 이용하여 이번엔 jupyter에서 자동으로 매겨준 index를 활용해 instance들을 삭제해보자.<br>
train.index를 입력해 train set의 index를 삭제한 정보를 valid set에 할당한다.

In [12]:
valid = titanic.drop(index=train.index)
valid

Survived  Pclass     Sex    Age  SibSp  Parch      Fare Embarked
1           1       1  female  38.00      1      0   71.2833        C
15          1       2  female  55.00      0      0   16.0000        S
20          0       2    male  35.00      0      0   26.0000        S
21          1       2    male  34.00      0      0   13.0000        S
22          1       3  female  15.00      0      0    8.0292        Q
27          0       1    male  19.00      3      2  263.0000        S
40          0       3  female  40.00      1      0    9.4750        S
49          0       3  female  18.00      1      0   17.8000        S
51          0       3    male  21.00      0      0    7.8000        S
56          1       2  female  21.00      0      0   10.5000        S
59          0       3    male  11.00      5      2   46.9000        S
60          0       3    male  22.00      0      0    7.2292        C
72          0       2    male  21.00      0      0   73.5000        S
73          0       3    male  26.00      1      0   14.4542        C
78          1       2    male   0.83      0      2   29.0000        S
84          1       2  female  17.00      0      0   10.5000        S
85          1       3  female  33.00      3      0   15.8500        S
86          0       3    male  16.00      1      3   34.3750        S
88          1       1  female  23.00      3      2  263.0000        S
89          0       3    male  24.00      0      0    8.0500        S
90          0       3    male  29.00      0      0    8.0500        S
91          0       3    male  20.00      0      0    7.8542        S
113         0       3  female  20.00      1      0    9.8250        S
119         0       3  female   2.00      4      2   31.2750        S
125         1       3    male  12.00      1      0   11.2417        C
141         1       3  female  22.00      0      0    7.7500        S
142         1       3  female  24.00      1      0   15.8500        S
147         0       3  female   9.00      2      2   34.3750        S
155         0       1    male  51.00      0      1   61.3792        C
162         0       3    male  26.00      0      0    7.7750        S
..        ...     ...     ...    ...    ...    ...       ...      ...
724         1       1    male  27.00      1      0   53.1000        S
725         0       3    male  20.00      0      0    8.6625        S
730         1       1  female  29.00      0      0  211.3375        S
733         0       2    male  23.00      0      0   13.0000        S
747         1       2  female  30.00      0      0   13.0000        S
748         0       1    male  19.00      1      0   53.1000        S
759         1       1  female  33.00      0      0   86.5000        S
767         0       3  female  30.50      0      0    7.7500        Q
774         1       2  female  54.00      1      3   23.0000        S
779         1       1  female  43.00      0      1  211.3375        S
780         1       3  female  13.00      0      0    7.2292        C
784         0       3    male  25.00      0      0    7.0500        S
787         0       3    male   8.00      4      1   29.1250        Q
788         1       3    male   1.00      1      2   20.5750        S
797         1       3  female  31.00      0      0    8.6833        S
801         1       2  female  31.00      1      1   26.2500        S
806         0       1    male  39.00      0      0    0.0000        S
811         0       3    male  39.00      0      0   24.1500        S
824         0       3    male   2.00      4      1   39.6875        S
830         1       3  female  15.00      1      0   14.4542        C
831         1       2    male   0.83      1      1   18.7500        S
843         0       3    male  34.50      0      0    6.4375        C
847         0       3    male  35.00      0      0    7.8958        C
857         1       1    male  51.00      0      0   26.5500        S
860         0       3    male  41.00      2      0   14.1083        S
864         0  

이제 학습의 target이 될 Survived 행을 따로 분리한다.<br>
loc은 label을 통해 행과 열에 접근하는 방식을 사용한다.<br>
comma 앞은 행을, 뒤는 열을 의미한다.
Survived는 맨 앞에 위치한 열이므로 전체 행에 대하여 Pclass부터 Embarked까지의 열을 train_x에 넣어주고, Survived 열만 따로 train_y에 넣어준다.

In [31]:
train_x = train.loc[:, 'Pclass':'Embarked']

In [32]:
train_x

Pclass     Sex    Age  SibSp  Parch      Fare Embarked
110       1    male  47.00      0      0   52.0000        S
814       3    male  30.50      0      0    8.0500        S
328       3  female  31.00      1      1   20.5250        S
882       3  female  22.00      0      0   10.5167        S
582       2    male  54.00      0      0   26.0000        S
569       3    male  32.00      0      0    7.8542        S
399       2  female  28.00      0      0   12.6500        S
489       3    male   9.00      1      1   15.9000        S
586       2    male  47.00      0      0   15.0000        S
434       1    male  50.00      1      0   55.9000        S
838       3    male  32.00      0      0   56.4958        S
590       3    male  35.00      0      0    7.1250        S
377       1    male  27.00      0      2  211.5000        C
670       2  female  40.00      1      1   39.0000        S
305       1    male   0.92      1      2  151.5500        S
329       1  female  16.00      0      1   57.9792        C
311       1  female  18.00      2      2  262.3750        C
133       2  female  29.00      1      0   26.0000        S
606       3    male  30.00      0      0    7.8958        S
664       3    male  20.00      1      0    7.9250        S
106       3  female  21.00      0      0    7.6500        S
414       3    male  44.00      0      0    7.9250        S
856       1  female  45.00      1      1  164.8667        S
74        3    male  32.00      0      0   56.4958        S
446       2  female  13.00      0      1   19.5000        S
58        2  female   5.00      1      2   27.7500        S
157       3    male  30.00      0      0    8.0500        S
14        3  female  14.00      0      0    7.8542        S
515       1    male  47.00      0      0   34.0208        S
12        3    male  20.00      0      0    8.0500        S
..      ...     ...    ...    ...    ...       ...      ...
679       1    male  36.00      0      1  512.3292        C
234       2    male  24.00      0      0   10.5000        S
238       2    male  19.00      0      0   10.5000        S
443       2  female  28.00      0      0   13.0000        S
875       3  female  15.00      0      0    7.2250        C
865       2  female  42.00      0      0   13.0000        S
758       3    male  34.00      0      0    8.0500        S
597       3    male  49.00      0      0    0.0000        S
871       1  female  47.00      1      1   52.5542        S
471       3    male  38.00      0      0    8.6625        S
528       3    male  39.00      0      0    7.9250        S
286       3    male  30.00      0      0    9.5000        S
224       1    male  38.00      1      0   90.0000        S
188       3    male  40.00      1      1   15.5000        Q
242       2    male  29.00      0      0   10.5000        S
809       1  female  33.00      1      0   53.1000        S
39        3  female  14.00      1      0   11.2417        C
2         3  female  26.00      0      0    7.9250        S
280       3    male  65.00      0      0    7.7500        Q
890       3    male  32.00      0      0    7.7500        Q
447       1    male  34.00      0      0   26.5500        S
233       3  female   5.00      4      2   31.3875        S
544       1    male  50.00      1      0  106.4250        C
81        3    male  29.00      0      0    9.5000        S
673       2    male  31.00      0      0   13.0000        S
858       3  female  24.00      0      3   19.2583        C
623       3    male  21.00      0      0    7.8542        S
706       2  female  45.00      0      0   13.5000        S
772       2  female  57.00      0      0   10.5000        S
777       3  female   5.00      0      0   12.4750        S

[570 rows x 7 columns]

In [33]:
train_y = train[['Survived']]

In [34]:
train_y

Survived
110         0
814         0
328         1
882         0
582         0
569         1
399         1
489         1
586         0
434         0
838         1
590         0
377         0
670         1
305         1
329         1
311         1
133         1
606         0
664         1
106         1
414         1
856         1
74          1
446         1
58          1
157         0
14          0
515         0
12          0
..        ...
679         1
234         0
238         0
443         1
875         1
865         1
758         0
597         0
871         1
471         0
528         0
286         1
224         1
188         0
242         0
809         1
39          1
2           1
280         0
890         0
447         1
233         1
544         0
81          1
673         1
858         1
623         0
706         1
772         0
777         1

[570 rows x 1 columns]

다음으로 valid set도 똑같이 분리시켜준다.

In [35]:
valid_x = valid.loc[:, 'Pclass':'Embarked']

In [36]:
valid_x

Pclass     Sex    Age  SibSp  Parch      Fare Embarked
1         1  female  38.00      1      0   71.2833        C
15        2  female  55.00      0      0   16.0000        S
20        2    male  35.00      0      0   26.0000        S
21        2    male  34.00      0      0   13.0000        S
22        3  female  15.00      0      0    8.0292        Q
27        1    male  19.00      3      2  263.0000        S
40        3  female  40.00      1      0    9.4750        S
49        3  female  18.00      1      0   17.8000        S
51        3    male  21.00      0      0    7.8000        S
56        2  female  21.00      0      0   10.5000        S
59        3    male  11.00      5      2   46.9000        S
60        3    male  22.00      0      0    7.2292        C
72        2    male  21.00      0      0   73.5000        S
73        3    male  26.00      1      0   14.4542        C
78        2    male   0.83      0      2   29.0000        S
84        2  female  17.00      0      0   10.5000        S
85        3  female  33.00      3      0   15.8500        S
86        3    male  16.00      1      3   34.3750        S
88        1  female  23.00      3      2  263.0000        S
89        3    male  24.00      0      0    8.0500        S
90        3    male  29.00      0      0    8.0500        S
91        3    male  20.00      0      0    7.8542        S
113       3  female  20.00      1      0    9.8250        S
119       3  female   2.00      4      2   31.2750        S
125       3    male  12.00      1      0   11.2417        C
141       3  female  22.00      0      0    7.7500        S
142       3  female  24.00      1      0   15.8500        S
147       3  female   9.00      2      2   34.3750        S
155       1    male  51.00      0      1   61.3792        C
162       3    male  26.00      0      0    7.7750        S
..      ...     ...    ...    ...    ...       ...      ...
724       1    male  27.00      1      0   53.1000        S
725       3    male  20.00      0      0    8.6625        S
730       1  female  29.00      0      0  211.3375        S
733       2    male  23.00      0      0   13.0000        S
747       2  female  30.00      0      0   13.0000        S
748       1    male  19.00      1      0   53.1000        S
759       1  female  33.00      0      0   86.5000        S
767       3  female  30.50      0      0    7.7500        Q
774       2  female  54.00      1      3   23.0000        S
779       1  female  43.00      0      1  211.3375        S
780       3  female  13.00      0      0    7.2292        C
784       3    male  25.00      0      0    7.0500        S
787       3    male   8.00      4      1   29.1250        Q
788       3    male   1.00      1      2   20.5750        S
797       3  female  31.00      0      0    8.6833        S
801       2  female  31.00      1      1   26.2500        S
806       1    male  39.00      0      0    0.0000        S
811       3    male  39.00      0      0   24.1500        S
824       3    male   2.00      4      1   39.6875        S
830       3  female  15.00      1      0   14.4542        C
831       2    male   0.83      1      1   18.7500        S
843       3    male  34.50      0      0    6.4375        C
847       3    male  35.00      0      0    7.8958        C
857       1    male  51.00      0      0   26.5500        S
860       3    male  41.00      2      0   14.1083        S
864       2    male  24.00      0      0   13.0000        S
873       3    male  47.00      0      0    9.0000        S
874       2  female  28.00      1      0   24.0000        C
876       3    male  20.00      0      0    9.8458        S
886       2    male  27.00      0      0   13.0000        S

[142 rows x 7 columns]

In [37]:
valid_y = valid[['Survived']]

In [38]:
valid_y

Survived
1           1
15          1
20          0
21          1
22          1
27          0
40          0
49          0
51          0
56          1
59          0
60          0
72          0
73          0
78          1
84          1
85          1
86          0
88          1
89          0
90          0
91          0
113         0
119         0
125         1
141         1
142         1
147         0
155         0
162         0
..        ...
724         1
725         0
730         1
733         0
747         1
748         0
759         1
767         0
774         1
779         1
780         1
784         0
787         0
788         1
797         1
801         1
806         0
811         0
824         0
830         1
831         1
843         0
847         0
857         1
860         0
864         0
873         0
874         1
876         0
886         0

[142 rows x 1 columns]

***
## 전처리
sklearn을 이용하여 데이터 분석 이전에 전처리를 시행한다.<br>
sklearn.preprocessing에 다양한 전처리 기법들이 존재한다.<br>
먼저 nominal data인 Sex와 Embarked를 numeric data로 바꿔준다.<br>
두 가지 방법이 존재하는데, OneHotEncoder와 OrdinalEncoder이다.

In [21]:
from sklearn.preprocessing import OneHotEncoder, OrdinalEncoder

OrdinalEncoder는 각 값에 대해 0부터 숫자를 부여하는 방식이다.<br>
Sex의 경우 Female = 0, Male = 1과 같은 식으로 인코딩될 수 있다.<br>
fit(): Train the model using libsvm

In [22]:
ord_enc = OrdinalEncoder()

In [23]:
ord_enc.fit(train_x[['Sex', 'Embarked']])

OrdinalEncoder(categories='auto', dtype=<class 'numpy.float64'>)

학습시킨 ord_enc가 정보를 어떻게 분류했는지 categories_를 통해 살펴볼 수 있다.<br>
Sex의 경우 female, male 순으로, Embarked의 경우 C,Q,S 순으로 분류되었다.<br>
즉, female에 0이, male에 1이 대입되었다는 말이다.

In [24]:
ord_enc.categories_

[array(['female', 'male'], dtype=object), array(['C', 'Q', 'S'], dtype=object)]

In [25]:
train_x[['Sex', 'Embarked']] = ord_enc.transform(train_x[['Sex', 'Embarked']])

In [26]:
train_x

Pclass  Sex    Age  SibSp  Parch      Fare  Embarked
110       1  1.0  47.00      0      0   52.0000       2.0
814       3  1.0  30.50      0      0    8.0500       2.0
328       3  0.0  31.00      1      1   20.5250       2.0
882       3  0.0  22.00      0      0   10.5167       2.0
582       2  1.0  54.00      0      0   26.0000       2.0
569       3  1.0  32.00      0      0    7.8542       2.0
399       2  0.0  28.00      0      0   12.6500       2.0
489       3  1.0   9.00      1      1   15.9000       2.0
586       2  1.0  47.00      0      0   15.0000       2.0
434       1  1.0  50.00      1      0   55.9000       2.0
838       3  1.0  32.00      0      0   56.4958       2.0
590       3  1.0  35.00      0      0    7.1250       2.0
377       1  1.0  27.00      0      2  211.5000       0.0
670       2  0.0  40.00      1      1   39.0000       2.0
305       1  1.0   0.92      1      2  151.5500       2.0
329       1  0.0  16.00      0      1   57.9792       0.0
311       1  0.0  18.00      2      2  262.3750       0.0
133       2  0.0  29.00      1      0   26.0000       2.0
606       3  1.0  30.00      0      0    7.8958       2.0
664       3  1.0  20.00      1      0    7.9250       2.0
106       3  0.0  21.00      0      0    7.6500       2.0
414       3  1.0  44.00      0      0    7.9250       2.0
856       1  0.0  45.00      1      1  164.8667       2.0
74        3  1.0  32.00      0      0   56.4958       2.0
446       2  0.0  13.00      0      1   19.5000       2.0
58        2  0.0   5.00      1      2   27.7500       2.0
157       3  1.0  30.00      0      0    8.0500       2.0
14        3  0.0  14.00      0      0    7.8542       2.0
515       1  1.0  47.00      0      0   34.0208       2.0
12        3  1.0  20.00      0      0    8.0500       2.0
..      ...  ...    ...    ...    ...       ...       ...
679       1  1.0  36.00      0      1  512.3292       0.0
234       2  1.0  24.00      0      0   10.5000       2.0
238       2  1.0  19.00      0      0   10.5000       2.0
443       2  0.0  28.00      0      0   13.0000       2.0
875       3  0.0  15.00      0      0    7.2250       0.0
865       2  0.0  42.00      0      0   13.0000       2.0
758       3  1.0  34.00      0      0    8.0500       2.0
597       3  1.0  49.00      0      0    0.0000       2.0
871       1  0.0  47.00      1      1   52.5542       2.0
471       3  1.0  38.00      0      0    8.6625       2.0
528       3  1.0  39.00      0      0    7.9250       2.0
286       3  1.0  30.00      0      0    9.5000       2.0
224       1  1.0  38.00      1      0   90.0000       2.0
188       3  1.0  40.00      1      1   15.5000       1.0
242       2  1.0  29.00      0      0   10.5000       2.0
809       1  0.0  33.00      1      0   53.1000       2.0
39        3  0.0  14.00      1      0   11.2417       0.0
2         3  0.0  26.00      0      0    7.9250       2.0
280       3  1.0  65.00      0      0    7.7500       1.0
890       3  1.0  32.00      0      0    7.7500       1.0
447       1  1.0  34.00      0      0   26.5500       2.0
233       3  0.0   5.00      4      2   31.3875       2.0
544       1  1.0  50.00      1      0  106.4250       0.0
81        3  1.0  29.00      0      0    9.5000       2.0
673       2  1.0  31.00      0      0   13.0000       2.0
858       3  0.0  24.00      0      3   19.2583       0.0
623       3  1.0  21.00      0      0    7.8542       2.0
706       2  0.0  45.00      0      0   13.5000       2.0
772       2  0.0  57.00      0      0   10.5000       2.0
777       3  0.0   5.00      0      0   12.4750       2.0

[570 rows x 7 columns]

다음은 valid set에도 똑같이 적용해보자.

In [27]:
ord_enc.fit(valid_x[['Sex', 'Embarked']])

OrdinalEncoder(categories='auto', dtype=<class 'numpy.float64'>)

In [28]:
ord_enc.categories_

[array(['female', 'male'], dtype=object), array(['C', 'Q', 'S'], dtype=object)]

그런데 train과 valid가 random하게 나뉘어졌기 때문에 category가 다르게 지정이 될 수 있다.<br>
현재는 같지만 다를 슈도 있다구.<br>
그러므로 OrdinalEncoder의 카테고리를 수동으로 지정해주자.

In [29]:
OrdinalEncoder(categories=[['male', 'female'],['C', 'S', 'Q']])

OrdinalEncoder(categories=[['male', 'female'], ['C', 'S', 'Q']],
        dtype=<class 'numpy.float64'>)

이번에는 OneHotEncoder를 이용해보자.<br>
OneHotEncoder는 각 카테고리의 수 만큼 행렬을 만들어 1 또는 0을 부여하는 방식이다.<br>
예를 들어, Sex의 경우 (male, female)이라는 일종의 벡터 형태를 지정해 값이 male인 경우 (1,0)으로 기술하는 식이다.<br>
반대로 값이 female인 경우에는 (0,1)이 될 것이다.

OneHotEncoder는 Sex와 Embarked에 따로 적용시켜주자.<br>
categories에 변환할 카테고리를 직접 지정해주면 된다.

In [30]:
onehot_sex_enc = OneHotEncoder(categories=[['male', 'female']])

fit_transform은 fit과 transform을 한꺼번에 해 줄 수 있다.
toarray()를 이용해 array형태로 만들어줄 수 있다.

In [39]:
onehot_sex_enc.fit_transform(train_x[['Sex']]).toarray()

array([[1., 0.],
       [1., 0.],
       [0., 1.],
       ...,
       [0., 1.],
       [0., 1.],
       [0., 1.]])

In [40]:
onehot_sex_enc.categories_

[array(['male', 'female'], dtype=object)]

이제 train_x 데이터에 변환한 Sex를 덮어씌워야 하는데, 에러가 난다.<br>
그 이유는 하나였던 Sex column이 OneHot을 이용하면서 female, male의 두 개의 column으로 늘어났기 때문이다.

In [41]:
train_x[['Sex']] = onehot_sex_enc.fit_transform(train_x[['Sex']]).toarray()

ValueError: Must have equal len keys and value when setting with an ndarray

먼저 index가 뒤죽박죽이므로 0부터 다시 매겨준다. reset_index()를 사용하면 됨.<br>
그런데 이걸 쓰면 index라는 열이 새로 생기므로 drop시켜준다.

In [42]:
train_x = train_x.reset_index()

In [43]:
train_x = train_x.drop(columns=['index'])

In [44]:
train_x

Pclass     Sex    Age  SibSp  Parch      Fare Embarked
0         1    male  47.00      0      0   52.0000        S
1         3    male  30.50      0      0    8.0500        S
2         3  female  31.00      1      1   20.5250        S
3         3  female  22.00      0      0   10.5167        S
4         2    male  54.00      0      0   26.0000        S
5         3    male  32.00      0      0    7.8542        S
6         2  female  28.00      0      0   12.6500        S
7         3    male   9.00      1      1   15.9000        S
8         2    male  47.00      0      0   15.0000        S
9         1    male  50.00      1      0   55.9000        S
10        3    male  32.00      0      0   56.4958        S
11        3    male  35.00      0      0    7.1250        S
12        1    male  27.00      0      2  211.5000        C
13        2  female  40.00      1      1   39.0000        S
14        1    male   0.92      1      2  151.5500        S
15        1  female  16.00      0      1   57.9792        C
16        1  female  18.00      2      2  262.3750        C
17        2  female  29.00      1      0   26.0000        S
18        3    male  30.00      0      0    7.8958        S
19        3    male  20.00      1      0    7.9250        S
20        3  female  21.00      0      0    7.6500        S
21        3    male  44.00      0      0    7.9250        S
22        1  female  45.00      1      1  164.8667        S
23        3    male  32.00      0      0   56.4958        S
24        2  female  13.00      0      1   19.5000        S
25        2  female   5.00      1      2   27.7500        S
26        3    male  30.00      0      0    8.0500        S
27        3  female  14.00      0      0    7.8542        S
28        1    male  47.00      0      0   34.0208        S
29        3    male  20.00      0      0    8.0500        S
..      ...     ...    ...    ...    ...       ...      ...
540       1    male  36.00      0      1  512.3292        C
541       2    male  24.00      0      0   10.5000        S
542       2    male  19.00      0      0   10.5000        S
543       2  female  28.00      0      0   13.0000        S
544       3  female  15.00      0      0    7.2250        C
545       2  female  42.00      0      0   13.0000        S
546       3    male  34.00      0      0    8.0500        S
547       3    male  49.00      0      0    0.0000        S
548       1  female  47.00      1      1   52.5542        S
549       3    male  38.00      0      0    8.6625        S
550       3    male  39.00      0      0    7.9250        S
551       3    male  30.00      0      0    9.5000        S
552       1    male  38.00      1      0   90.0000        S
553       3    male  40.00      1      1   15.5000        Q
554       2    male  29.00      0      0   10.5000        S
555       1  female  33.00      1      0   53.1000        S
556       3  female  14.00      1      0   11.2417        C
557       3  female  26.00      0      0    7.9250        S
558       3    male  65.00      0      0    7.7500        Q
559       3    male  32.00      0      0    7.7500        Q
560       1    male  34.00      0      0   26.5500        S
561       3  female   5.00      4      2   31.3875        S
562       1    male  50.00      1      0  106.4250        C
563       3    male  29.00      0      0    9.5000        S
564       2    male  31.00      0      0   13.0000        S
565       3  female  24.00      0      3   19.2583        C
566       3    male  21.00      0      0    7.8542        S
567       2  female  45.00      0      0   13.5000        S
568       2  female  57.00      0      0   10.5000        S
569       3  female   5.00      0      0   12.4750        S

[570 rows x 7 columns]

sex_df라는 dataset을 따로 만든 뒤에 기존의 train_x에 있던 Sex를 빼고 넣어준다.

In [45]:
sex_df = pd.DataFrame(data=onehot_sex_enc.fit_transform(train_x[['Sex']]).toarray(),
                      columns=['male', 'female'])

In [46]:
sex_df

male  female
0     1.0     0.0
1     1.0     0.0
2     0.0     1.0
3     0.0     1.0
4     1.0     0.0
5     1.0     0.0
6     0.0     1.0
7     1.0     0.0
8     1.0     0.0
9     1.0     0.0
10    1.0     0.0
11    1.0     0.0
12    1.0     0.0
13    0.0     1.0
14    1.0     0.0
15    0.0     1.0
16    0.0     1.0
17    0.0     1.0
18    1.0     0.0
19    1.0     0.0
20    0.0     1.0
21    1.0     0.0
22    0.0     1.0
23    1.0     0.0
24    0.0     1.0
25    0.0     1.0
26    1.0     0.0
27    0.0     1.0
28    1.0     0.0
29    1.0     0.0
..    ...     ...
540   1.0     0.0
541   1.0     0.0
542   1.0     0.0
543   0.0     1.0
544   0.0     1.0
545   0.0     1.0
546   1.0     0.0
547   1.0     0.0
548   0.0     1.0
549   1.0     0.0
550   1.0     0.0
551   1.0     0.0
552   1.0     0.0
553   1.0     0.0
554   1.0     0.0
555   0.0     1.0
556   0.0     1.0
557   0.0     1.0
558   1.0     0.0
559   1.0     0.0
560   1.0     0.0
561   0.0     1.0
562   1.0     0.0
563   1.0     0.0
564   1.0     0.0
565   0.0     1.0
566   1.0     0.0
567   0.0     1.0
568   0.0     1.0
569   0.0     1.0

[570 rows x 2 columns]

In [47]:
train_x = pd.concat((train_x[['Pclass']], sex_df, train_x.loc[:, 'Age':'Embarked']), axis = 1)

In [49]:
train_x

Pclass  male  female    Age  SibSp  Parch      Fare Embarked
0         1   1.0     0.0  47.00      0      0   52.0000        S
1         3   1.0     0.0  30.50      0      0    8.0500        S
2         3   0.0     1.0  31.00      1      1   20.5250        S
3         3   0.0     1.0  22.00      0      0   10.5167        S
4         2   1.0     0.0  54.00      0      0   26.0000        S
5         3   1.0     0.0  32.00      0      0    7.8542        S
6         2   0.0     1.0  28.00      0      0   12.6500        S
7         3   1.0     0.0   9.00      1      1   15.9000        S
8         2   1.0     0.0  47.00      0      0   15.0000        S
9         1   1.0     0.0  50.00      1      0   55.9000        S
10        3   1.0     0.0  32.00      0      0   56.4958        S
11        3   1.0     0.0  35.00      0      0    7.1250        S
12        1   1.0     0.0  27.00      0      2  211.5000        C
13        2   0.0     1.0  40.00      1      1   39.0000        S
14        1   1.0     0.0   0.92      1      2  151.5500        S
15        1   0.0     1.0  16.00      0      1   57.9792        C
16        1   0.0     1.0  18.00      2      2  262.3750        C
17        2   0.0     1.0  29.00      1      0   26.0000        S
18        3   1.0     0.0  30.00      0      0    7.8958        S
19        3   1.0     0.0  20.00      1      0    7.9250        S
20        3   0.0     1.0  21.00      0      0    7.6500        S
21        3   1.0     0.0  44.00      0      0    7.9250        S
22        1   0.0     1.0  45.00      1      1  164.8667        S
23        3   1.0     0.0  32.00      0      0   56.4958        S
24        2   0.0     1.0  13.00      0      1   19.5000        S
25        2   0.0     1.0   5.00      1      2   27.7500        S
26        3   1.0     0.0  30.00      0      0    8.0500        S
27        3   0.0     1.0  14.00      0      0    7.8542        S
28        1   1.0     0.0  47.00      0      0   34.0208        S
29        3   1.0     0.0  20.00      0      0    8.0500        S
..      ...   ...     ...    ...    ...    ...       ...      ...
540       1   1.0     0.0  36.00      0      1  512.3292        C
541       2   1.0     0.0  24.00      0      0   10.5000        S
542       2   1.0     0.0  19.00      0      0   10.5000        S
543       2   0.0     1.0  28.00      0      0   13.0000        S
544       3   0.0     1.0  15.00      0      0    7.2250        C
545       2   0.0     1.0  42.00      0      0   13.0000        S
546       3   1.0     0.0  34.00      0      0    8.0500        S
547       3   1.0     0.0  49.00      0      0    0.0000        S
548       1   0.0     1.0  47.00      1      1   52.5542        S
549       3   1.0     0.0  38.00      0      0    8.6625        S
550       3   1.0     0.0  39.00      0      0    7.9250        S
551       3   1.0     0.0  30.00      0      0    9.5000        S
552       1   1.0     0.0  38.00      1      0   90.0000        S
553       3   1.0     0.0  40.00      1      1   15.5000        Q
554       2   1.0     0.0  29.00      0      0   10.5000        S
555       1   0.0     1.0  33.00      1      0   53.1000        S
556       3   0.0     1.0  14.00      1      0   11.2417        C
557       3   0.0     1.0  26.00      0      0    7.9250        S
558       3   1.0     0.0  65.00      0      0    7.7500        Q
559       3   1.0     0.0  32.00      0      0    7.7500        Q
560       1   1.0     0.0  34.00      0      0   26.5500        S
561       3   0.0     1.0   5.00      4      2   31.3875        S
562       1   1.0     0.0  50.00      1      0  106.4250        C
563       3   1.0     0.0  29.00      0      0    9.5000        S
564       2   1.0     0.0  31.00      0      0   13.0000        S
565       3   0.0     1.0  24.00      0      3   19.2583        C
566       3   1.0     0.0  21.00      0      0    7.8542        S
567       2   0.0     1.0  45.00      0      0   13.5000        S
568       2   0.0     1.0  57.00      0      

Embarked에 대해서도 동일하게 진행한다.

In [50]:
onehot_emb_enc = OneHotEncoder(categories=[['C','S','Q']])

In [51]:
onehot_emb_enc.fit(train_x[['Embarked']])

OneHotEncoder(categorical_features=None, categories=[['C', 'S', 'Q']],
       dtype=<class 'numpy.float64'>, handle_unknown='error',
       n_values=None, sparse=True)

In [52]:
onehot_emb_enc.transform(train_x[['Embarked']]).toarray()

array([[0., 1., 0.],
       [0., 1., 0.],
       [0., 1., 0.],
       ...,
       [0., 1., 0.],
       [0., 1., 0.],
       [0., 1., 0.]])

In [53]:
emb_df = pd.DataFrame(data=onehot_emb_enc.transform(train_x[['Embarked']]).toarray(),
                     columns = ['Cherbourg', 'Southampton', 'Q'])

In [58]:
train_x = pd.concat((train_x.drop(columns = 'Embarked'), emb_df), axis=1)

In [59]:
train_x

Pclass  male  female    Age  SibSp  Parch      Fare  Cherbourg  \
0         1   1.0     0.0  47.00      0      0   52.0000        0.0   
1         3   1.0     0.0  30.50      0      0    8.0500        0.0   
2         3   0.0     1.0  31.00      1      1   20.5250        0.0   
3         3   0.0     1.0  22.00      0      0   10.5167        0.0   
4         2   1.0     0.0  54.00      0      0   26.0000        0.0   
5         3   1.0     0.0  32.00      0      0    7.8542        0.0   
6         2   0.0     1.0  28.00      0      0   12.6500        0.0   
7         3   1.0     0.0   9.00      1      1   15.9000        0.0   
8         2   1.0     0.0  47.00      0      0   15.0000        0.0   
9         1   1.0     0.0  50.00      1      0   55.9000        0.0   
10        3   1.0     0.0  32.00      0      0   56.4958        0.0   
11        3   1.0     0.0  35.00      0      0    7.1250        0.0   
12        1   1.0     0.0  27.00      0      2  211.5000        1.0   
13        2   0.0     1.0  40.00      1      1   39.0000        0.0   
14        1   1.0     0.0   0.92      1      2  151.5500        0.0   
15        1   0.0     1.0  16.00      0      1   57.9792        1.0   
16        1   0.0     1.0  18.00      2      2  262.3750        1.0   
17        2   0.0     1.0  29.00      1      0   26.0000        0.0   
18        3   1.0     0.0  30.00      0      0    7.8958        0.0   
19        3   1.0     0.0  20.00      1      0    7.9250        0.0   
20        3   0.0     1.0  21.00      0      0    7.6500        0.0   
21        3   1.0     0.0  44.00      0      0    7.9250        0.0   
22        1   0.0     1.0  45.00      1      1  164.8667        0.0   
23        3   1.0     0.0  32.00      0      0   56.4958        0.0   
24        2   0.0     1.0  13.00      0      1   19.5000        0.0   
25        2   0.0     1.0   5.00      1      2   27.7500        0.0   
26        3   1.0     0.0  30.00      0      0    8.0500        0.0   
27        3   0.0     1.0  14.00      0      0    7.8542        0.0   
28        1   1.0     0.0  47.00      0      0   34.0208        0.0   
29        3   1.0     0.0  20.00      0      0    8.0500        0.0   
..      ...   ...     ...    ...    ...    ...       ...        ...   
540       1   1.0     0.0  36.00      0      1  512.3292        1.0   
541       2   1.0     0.0  24.00      0      0   10.5000        0.0   
542       2   1.0     0.0  19.00      0      0   10.5000        0.0   
543       2   0.0     1.0  28.00      0      0   13.0000        0.0   
544       3   0.0     1.0  15.00      0      0    7.2250        1.0   
545       2   0.0     1.0  42.00      0      0   13.0000        0.0   
546       3   1.0     0.0  34.00      0      0    8.0500        0.0   
547       3   1.0     0.0  49.00      0      0    0.0000        0.0   
548       1   0.0     1.0  47.00      1      1   52.5542        0.0   
549       3   1.0     0.0  38.00      0      0    8.6625        0.0   
550       3   1.0     0.0  39.00      0      0    7.9250        0.0   
551       3   1.0     0.0  30.00      0      0    9.5000        0.0   
552       1   1.0     0.0  38.00      1      0   90.0000        0.0   
553       3   1.0     0.0  40.00      1      1   15.5000        0.0   
554       2   1.0     0.0  29.00      0      0   10.5000        0.0   
555       1   0.0     1.0  33.00      1      0   53.1000        0.0   
556       3   0.0     1.0  14.00      1      0   11.2417        1.0   
557       3   0.0     1.0  26.00      0      0    7.9250        0.0   
558       3   1.0     0.0  65.00      0      0    7.7500        0.0   
559       3   1.0     0.0  32.00      0      0    7.7500        0.0   
560       1   1.0     0.0  34.00      0      0   26.5500        0.0   
561       3   0.0     1.0   5.00      4      2   31.3875        0.0   
562       1   1.0     0.0  50.00      1      0  106.4250        1.0   
563       3   1.0     0.0  29.00      0      0    9.5000        0.0   
564       2   1.0     0.0  31

In [ ]:
from sklearn.preprocessing import StandardScaler

In [ ]:
StandardScaler()   #Scaler: 데이터를 표준분포를 따르도록 변형시켜 줌. Age와 Fare 같은 경우 표준정규분포로 바꾸면 이득.
#feature들 간 scaling이 다르다면 MLP의 학습에 문제가 있을 수 있음.
#MinMaxScaler는 scale의 범위 지정해주는 것.
#RobustScalaer는 outlier들의 값이 클 때 적용시켜주면 좋음.

In [ ]:
from sklearn.preprocessing import imputer  #missing value를 채워줌.

In [ ]:
from sklearn.preprocessing import Normalizer   #instance 단위의 scaling이라고 보면 됨.

In [ ]:
from sklearn.tree import DecisionTreeClassifier 

In [ ]:
from sklearn.naive_bayes import       #우리가 수업에서 했던 거는 성능이 넘 별로라 없음

In [ ]:
from sklearn.linear_model import LogisticRegression   #로지스틱은 사실 regression 모델이 아니라 classification이다~

In [ ]:
from sklearn.neural_network import MLPClassifier

In [ ]:
MLPClassifier(hidden_layer_sizes=(100,), activation='logistic', )   #keras에서 다시 설명하겠음